### PR TITLE
RFC for Server SDK NCCO Builders

### PR DESCRIPTION
**Pull request by hummusonrails**

### CODE REVIEW

File: rfcs/proposed/RFC-2-server-sdk-ncco-builders.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+We can offer a comprehensive NCCO builder in each of the mainline Vonage SDKs that will ease the NCCO action composition burden on our users considerably. These builders will assist users in both creating their NCCO actions, and provide real-time feedback on incorrect construction. 
+
+## Proposed Solution
+
+### Technical Specifications
+
+The NCCO builder MUST be instantiable. By providing distinct instantiations for each NCCO action, it becomes easier to provide specific feedback on each unique action for the developer. For example, there may be multiple `type` parameters throughout the NCCO actions, but the specifications for `type` will be different for each unique NCCO action. The disaggregation and unique instantiation of each action as its own class allows for that fine attention to detail.
```
Comment (dragonmantank) [inline rfcs/proposed/RFC-2-server-sdk-ncco-builders.md]: Splitting out that we should have a builder an individual actions that can be created. 

```suggestion
Each SDK MUST provide an NCCO Builder, which combines multiple actions into a singular NCCO collection, and NCCO Actions, which are individual actions that can be taken. The NCCO builder and NCCO actions MUST be instantiable. By providing distinct instantiations, it becomes easier to provide specific feedback on each unique action for the developer. For example, there may be multiple `type` parameters throughout the NCCO actions, but the specifications for `type` will be different for each unique NCCO action. The disaggregation and unique instantiation of each action as its own class allows for that fine attention to detail.
```

File: rfcs/proposed/RFC-2-server-sdk-ncco-builders.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+# Standalone NCCO with two actions
+
+talk = Vonage::Voice::Ncco.talk(text: 'Hello World!')
+input = Vonage::Voice::Ncco.input(type: ['dtmf'], dtmf: { bargeIn: true })
+ncco = Vonage::Voice::Ncco.build(talk, input)
+
+# => [{:action=>"talk", :text=>"Hello World!"}, {:action=>"input", :type=>["dtmf"], :dtmf=>{:bargeIn=>true}}]
+```
+
+Each NCCO action MUST be able to be accessed via its own method name accessed on the NCCO class. For example, the `connect` action would be accessed as `Vonage::Voice::Ncco.connect` and the `notify` action would be accessed as `Vonage::Voice::Ncco.notify`. For example, the following would accomplish this in the Ruby SDK:
```
Comment (dragonmantank) [inline rfcs/proposed/RFC-2-server-sdk-ncco-builders.md]: I would switch this from `MUST` to `MAY`. This reduces coupling between the builder class and the action classes, but still allows this structure as an option. It also helps sidestep the flexibility some languages provide where others don't or can't. 

```suggestion
Each NCCO action MAY be able to be accessed via its own method name accessed on the NCCO builder class. For example, the `connect` action would be accessed as `Vonage::Voice::Ncco.connect` and the `notify` action would be accessed as `Vonage::Voice::Ncco.notify`. For example, the following would accomplish this in the Ruby SDK:
```

File: rfcs/proposed/RFC-2-server-sdk-ncco-builders.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    end
+
+    def self.create(*actions)
+      actions.flatten!
+    end
+  end
+end
+```
+
+The NCCO builder MUST NOT be part of the SDK client instantiation. Adding the NCCO builder to the client could cause confusion for users as the expectation for client methods is a direct interaction with the Vonage APIs. In this regard, it SHOULD be similar to the JWT generator as specified in its [RFC](https://raw.githubusercontent.com/Nexmo/server-sdk-specification/main/rfcs/accepted/RFC-1-server-sdk-jwt-packages.md).
```
Comment (dragonmantank) [inline rfcs/proposed/RFC-2-server-sdk-ncco-builders.md]: I would reword this to state the NCCO object shouldn't depend on the main SDK client, as opposed to it shouldn't be part of client instantiation. I don't see a problem with getting the builder from the client god object, but there is no reason the NCCO objects should take a client as a dependency.

```suggestion
The NCCO builder MUST NOT require an SDK Client as a dependency. As the NCCO Builder does not access the Vonage APIs, it MUST be standalone objects. In this regard, it SHOULD be similar to the JWT generator as specified in its [RFC](https://raw.githubusercontent.com/Nexmo/server-sdk-specification/main/rfcs/accepted/RFC-1-server-sdk-jwt-packages.md).
```

File: rfcs/proposed/RFC-2-server-sdk-ncco-builders.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+The NCCO builder MUST return a fully constructed NCCO in the form of an array, with each action represented as a hash inside the array.
+
+### Field Specifications
```
Comment (dragonmantank) [inline rfcs/proposed/RFC-2-server-sdk-ncco-builders.md:77]: I would remove the "default" value. I found when building the PHP NCCO's that passing the default value can in many cases be different behavior that not sending any value. I would change the heading to "Values" if they act as enums or have ranges and list the values (like Record.format)

**Review by dragonmantank — CHANGES_REQUESTED** [review]
Some rewording, and suggestion for default values

File: rfcs/proposed/RFC-2-server-sdk-ncco-builders.md (hunk 4 above)
Comment (hummusonrails) [reply rfcs/proposed/RFC-2-server-sdk-ncco-builders.md:77]: That's a really good point. I'll get that pushed up today.

Comment (hummusonrails) [reply rfcs/proposed/RFC-2-server-sdk-ncco-builders.md:77]: @dragonmantank, done ✅

**Review by dragonmantank — APPROVED** [review]
LGTM to go to proposed
